tests/test_group.rs

```diff
@@ -322,106 +328,351 @@ fn group_operations() {
+            panic!("Different public trees");
+        }
 
-    group_alice_1234.process_commit(commit7.clone());
-    group_bob.process_commit(commit7.clone());
-    group_charlie.process_commit(commit7);
+        // === Bob adds Charlie ===
+        let charlie_credential_bundle =
+            CredentialBundle::new("Charlie".into(), CredentialType::Basic, ciphersuite.name())
+                .unwrap();
 
-    group_alice_1234.tree.print(&format!("\n{:?}", ms7));
+        let charlie_key_package_bundle = KeyPackageBundle::new(
+            &[ciphersuite.name()],
+            &charlie_credential_bundle,
+            mandatory_extensions.clone(),
+        )
+        .unwrap();
+        let charlie_key_package = charlie_key_package_bundle.get_key_package().clone();
 
-    // Alice updates
-    let update_proposal_alice = group_alice_1234.create_update_proposal(None);
+        let add_charlie_proposal_bob =
+            group_bob.create_add_proposal(&[], &bob_credential_bundle, charlie_key_package);
 
-    group_bob.process_proposal(update_proposal_alice.clone());
-    group_charlie.process_proposal(update_proposal_alice);
+        let (mls_plaintext_commit, welcome_for_charlie_option, kpb_option) = match group_bob
+            .create_commit(
+                &[],
+                &bob_credential_bundle,
+                vec![add_charlie_proposal_bob.clone()],
+                false, /* force self update*/
+            ) {
+            Ok(c) => c,
+            Err(e) => panic!("Error creating commit: {:?}", e),
+        };
 
-    let (commit8, ms8, _) = group_alice_1234.create_commit(None);
+        // Check there is no KeyPackageBundle since there are only Add Proposals and no forced self-update
+        assert!(kpb_option.is_none());
+        // Make sure the is a Welcome message for Charlie
+        assert!(welcome_for_charlie_option.is_some());
+
+        group_alice
+            .apply_commit(
+                mls_plaintext_commit.clone(),
+                vec![add_charlie_proposal_bob.clone()],
+                &[],
+            )
+            .expect("Error applying commit (Alice)");
+        group_bob
+            .apply_commit(
+                mls_plaintext_commit.clone(),
+                vec![add_charlie_proposal_bob],
+                &[],
+            )
+            .expect("Error applying commit (Bob)");
+
+        let ratchet_tree = group_alice.tree().public_key_tree_copy();
+        let mut group_charlie = match MlsGroup::new_from_welcome(
+            welcome_for_charlie_option.unwrap(),
+            Some(ratchet_tree),
+            charlie_key_package_bundle,
+        ) {
+            Ok(group) => group,
+            Err(e) => panic!("Error creating group from Welcome: {:?}", e),
+        };
 
-    group_alice_1234.process_commit(commit8.clone());
-    group_bob.process_commit(commit8.clone());
-    group_charlie.process_commit(commit8);
+        // Make sure that all groups have the same public tree
```

Comment:
Maybe it would be worth writing a `group_agreement` test, that takes a vector of group states and tests that they are equivalent in some ways, e.g. by comparing the public tree, encrypting/decrypting messages, etc?

Reply:
Sure, but any further testing it out of scope for this PR.

Reply:
It's not necessarily additional testing, it's just externalizing the "groups-are-equal" assertion into an additional function that does that for us.